### PR TITLE
New PBR techniques. New PBR terrain shader. Skybox shader fixed.

### PR DESCRIPTION
**Pull request by gleblebedev**
- New PBR techinques:
  Not all PBR materials have diffuse textures. It is a valid case to have only normal map defined as a texture and everything else though colors and scalar values. To match this expectation I've added technique permutations for the diffuseless materials.
- Skybox fixed:
  It is a bit inconvenient to have skybox node to be placed exactly at the zero position in the world space especially because the editor doesn't create it there by default. To have an option to ignore node transformation I've amended the shader. Also it looks like HLSL and GLSL shaders didn't match there.
- PBRTerrainBlend shader and technique:
  I've added PBRTerrainBlend to match terrain visually with the rest of the scene if it has models with PBR materials.

### CODE REVIEW

**Comment by gleblebedev** [comment]
@1vanK may I ask you to look at the changes please?

File: bin/CoreData/Shaders/GLSL/Skybox.glsl [hunk 1]
```diff
@@ -6,7 +6,13 @@ varying vec3 vTexCoord;
 
 void VS()
 {
+#ifdef IGNORENODETRANSFORM
+    mat4 m = transpose(cViewInv); 
+    m[0] = vec4(1,0,0,0); m[1] = vec4(0,1,0,0); m[2] = vec4(0,0,1,0);
```
Comment (1vanK) [inline bin/CoreData/Shaders/GLSL/Skybox.glsl:11]: Use 1.0 instead 1 to compatibility with old OpenGL https://github.com/1vanK/Urho3DOutline/pull/1
Also spaces afer ,

File: bin/CoreData/Shaders/HLSL/Skybox.hlsl [hunk 2]
```diff
@@ -9,10 +9,14 @@ void VS(float4 iPos : POSITION,
     out float3 oTexCoord : TEXCOORD0,
     out float4 oPos : OUTPOSITION)
 {
+#ifdef IGNORENODETRANSFORM
+    float4x3 m = cViewInv;
+    float4x3 modelMatrix = { float3(1,0,0), float3(0,1,0), float3(0,0,1), m[3] };
```
Comment (1vanK) [inline bin/CoreData/Shaders/HLSL/Skybox.hlsl:14]: Maybe just float4x3 modelMatrix = { float3(1,0,0), float3(0,1,0), float3(0,0,1), cViewInv[3] };

**Comment by 1vanK** [comment]
> editor doesn't create it there by default

Editor settings.. > New node mode = In center

**Comment by 1vanK** [comment]
I think we do not need an additional define IGNORENODETRANSFORM to fix the bug. But we still need use rotation from skybox, just ignore position.

**Comment by 1vanK** [comment]
Can provide a scene for test PBR terrain?

File: bin/CoreData/Shaders/GLSL/PBRTerrainBlend.glsl [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        #endif
+
+        vec3 toCamera = normalize(vWorldPos.xyz - cCameraPosPS);
+        vec3 reflection = normalize(reflect(toCamera, normal));
+
+        vec3 cubeColor = vVertexLight.rgb;
+
+        #ifdef IBL
+          vec3 iblColor = ImageBasedLighting(reflection, normal, toCamera, diffColor.rgb, specColor.rgb, roughness, cubeColor);
+          float gamma = 0.0;
```
Comment (1vanK) [inline bin/CoreData/Shaders/GLSL/PBRTerrainBlend.glsl:220]: not used?

File: bin/CoreData/Shaders/HLSL/PBRTerrainBlend.hlsl [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        #endif
+
+        const float3 toCamera = normalize(iWorldPos.xyz - cCameraPosPS);
+
+        const float3 reflection = normalize(reflect(toCamera, normal));
+        float3 cubeColor = iVertexLight.rgb;
+
+        #ifdef IBL
+            const float3 iblColor = ImageBasedLighting(reflection, normal, toCamera, diffColor, specColor, roughness, cubeColor);
+            const float gamma = 0;
```
Comment (1vanK) [inline bin/CoreData/Shaders/HLSL/PBRTerrainBlend.hlsl:274]: the same

File: bin/CoreData/Shaders/GLSL/PBRTerrainBlend.glsl (hunk 3 above)
Comment (gleblebedev) [reply bin/CoreData/Shaders/GLSL/PBRTerrainBlend.glsl:220]: I've applied same changes as Terrain blend has over LitSolid to PBRLitSolid. I'm quite sure the IBL is in use there.

Comment (gleblebedev) [reply bin/CoreData/Shaders/GLSL/PBRTerrainBlend.glsl:220]: As defined in thechique:
<technique vs="PBRTerrainBlend" ps="PBRTerrainBlend" psdefines="PBR IBL">

It is certainly used.

File: bin/CoreData/Shaders/HLSL/PBRTerrainBlend.hlsl (hunk 4 above)
Comment (gleblebedev) [reply bin/CoreData/Shaders/HLSL/PBRTerrainBlend.hlsl:274]: As defined in thechique:
<technique vs="PBRTerrainBlend" ps="PBRTerrainBlend" psdefines="PBR IBL">

It is certainly used.

**Comment by gleblebedev** [comment]
> the bug
I've check with author of the code ( @cadaver ) and according to him this isn't a bug. This takes into account node transform by design.

**Comment by gleblebedev** [comment]
> Can you provide a scene for test PBR terrain?

Done

**Comment by 1vanK** [comment]
Thanks!

**Comment by 1vanK** [comment]
PBRTerrain in this example wery bright. 

This can be fixed be adding files: `Data\Textures\TerrainDetail1.xml`, `Data\Textures\TerrainDetail2.xml`, `Data\Textures\TerrainDetail3.xml`, with content
```
<texture>
    <srgb enable="true" />
</texture>
```

But this broke another examples, used not PBR terrains (19_VehicleDemo for example).

I think we need to change PBRTerrain.xml to use PBR textures from `Autoload\LargeData\Textures\PBR`. 

```
<material>
    <technique name="Techniques/PBR/PBRTerrainBlend.xml" />
    <texture unit="0" name="Textures/TerrainWeights.dds" />
    <texture unit="1" name="Textures/PBR/Concrete/Albedo.jpg" />
    <texture unit="2" name="Textures/PBR/Mud/Albedo.jpg" />
    <texture unit="3" name="Textures/PBR/Sand/Albedo.jpg" />
    <parameter name="MatSpecColor" value="1.0 1.0 1.0 1.0" />
    <parameter name="DetailTiling" value="32 32" />
    <parameter name="Roughness" value="1" />
    <parameter name="Metallic" value="0" />
</material>
```

**Comment by gleblebedev** [comment]
Ok,will fix it today.

**Pull request merged**